exporters/otlp/internal/transform/metric.go

```diff
@@ -17,20 +17,210 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// source is complete, check for any errors.
+	if err := <-errc; err != nil {
+		return nil, err
+	}
+	return rms, nil
+}
+
+// source starts a goroutine that sends each one of the Records yielded by
+// the CheckpointSet on the returned chan. Any error encoutered will be sent
+// on the returned error chan after seeding is complete.
+func source(ctx context.Context, cps export.CheckpointSet) (<-chan export.Record, <-chan error) {
+	errc := make(chan error, 1)
+	out := make(chan export.Record)
+	// Seed records into process.
+	go func() {
+		defer close(out)
+		// No selected needed since errc is buffered.
+		errc <- cps.ForEach(func(r export.Record) error {
+			select {
+			case <-ctx.Done():
+				return ErrContextCanceled
+			case out <- r:
+			}
+			return nil
+		})
+	}()
+	return out, errc
+}
+
+// transformer transforms records read from the passed in chan into
+// OTLP Metrics which are sent on the out chan.
+func transformer(ctx context.Context, in <-chan export.Record, out chan<- result) {
+	for r := range in {
+		m, err := Record(r)
+		// Propagate errors, but do not send empty results.
+		if err == nil && m == nil {
+			continue
+		}
+		res := result{
+			Resource: r.Descriptor().Resource(),
+			Library:  r.Descriptor().LibraryName(),
+			Metric:   m,
+			Err:      err,
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case out <- res:
+		}
+	}
+}
+
+// sink collects transformed Records and batches them.
+//
+// Any errors encoutered transforming input will be reported with an
+// ErrTransforming as well as the completed ResourceMetrics. It is up to the
+// caller to handle any incorrect data in these ResourceMetrics.
+func sink(ctx context.Context, in <-chan result) ([]*metricpb.ResourceMetrics, error) {
+	var errStrings []string
+
+	type resourceBatch struct {
+		Resource *resourcepb.Resource
+		// Group by instrumentation library name and then the MetricDescriptor.
+		InstrumentationLibraryBatches map[string]map[string]*metricpb.Metric
+	}
+
+	// group by unique Resource string.
+	grouped := make(map[string]resourceBatch)
+	for res := range in {
+		if res.Err != nil {
+			errStrings = append(errStrings, res.Err.Error())
+			continue
+		}
+
+		rb, ok := grouped[res.Resource.String()]
+		if !ok {
+			rb = resourceBatch{
+				Resource:                      Resource(&res.Resource),
+				InstrumentationLibraryBatches: make(map[string]map[string]*metricpb.Metric),
+			}
+			grouped[res.Resource.String()] = rb
+		}
+
+		mb, ok := rb.InstrumentationLibraryBatches[res.Library]
+		if !ok {
+			mb = make(map[string]*metricpb.Metric)
+			rb.InstrumentationLibraryBatches[res.Library] = mb
+		}
+
+		m, ok := mb[res.Metric.GetMetricDescriptor().String()]
+		if !ok {
+			mb[res.Metric.GetMetricDescriptor().String()] = res.Metric
+			continue
+		}
+		if len(res.Metric.Int64DataPoints) > 0 {
+			m.Int64DataPoints = append(m.Int64DataPoints, res.Metric.Int64DataPoints...)
```

Comment:
Timestamp is not set anywhere for the dataPoints. Without timestamp multiple dataPoints will be ambiguous. 
Is there are a chance for two dataPoints to have the same timestamp? Probably not.

Reply:
This is a good point, and I think a more global problem for our Metrics pipeline. There isn't a way to know in our current structure when a metric instrument made a measurement and therefore no way to differentiate these data points.

I'm not sure how the collector handles these, but from the exporter perspective I wanted to transmit each separately instead of deciding on a way to combine them. I'm open to suggestions on different approaches and will add it to today's agenda for our SIG meeting to talk more about.

Reply:
Created https://github.com/open-telemetry/opentelemetry-go/issues/629 to track building support for this into the SDK.